### DISCUSSION TITLE
Alpine error handling

### DISCUSSION THREAD

**Discussion by personalsoft-fabianobonin** [alpinejs/alpine#3496] [7. Bugs (archived)]
Hi, guys.

In javascript we can intercept unhandled errors with "error" (sync) and with "unhandledrejection" (async) window events.

If you click the buttons below, you will see:

button 1 (sync, onclick) generating an "error" event
button 2 (async, onclick) generating an "unhandledrejection" event

button 3 (sync, @click) generating an "error" event
button 4 (async, @click) generating **BOTH** an "error" event and an "unhandledrejection" event

In this case, an application with global handlers for both "error" and "unhandledrejection" will handle the same error twice when they happen in an async process controled by Alpine, and it seems an inconsistent behavior.

Is it expected?

Regards,

Fabiano

```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <script src="//unpkg.com/alpinejs"
      defer></script>
  </head>
  <body x-data="{}">
    <button onclick="xxx()">button1 - js sync (will log "error")</button><br>
    <button onclick="yyy()">button2 - js async (will log "unhandledrejection")</button><br>
    <button @click="xxx()">button3 - alpine sync (will log "error")</button><br>
    <button @click="yyy()">button4 - alpine async (will log "error" AND "unhandledrejection")</button><br>

    <script>
      function xxx() {
        throw new Error("xxx");
      }

      async function yyy() {
        throw new Error("yyy");
      }

      // Handle errors
      window.addEventListener("error", (event) => {
        console.log("error");
      });

      // Handle rejected promises
      window.addEventListener("unhandledrejection", (event) => {
        console.log("unhandledrejection");
      });
    </script>
  </body>
</html>
```

**Comment by SimoTod** [comment]
all errors are caught and re-thrown in order to display useful debug warnings in the console. I think the rethrown will then count as a sync error,

**Reply by personalsoft-fabianobonin** [reply]
In `error.js`, when an error is handled in an async context, it is rethrown in a sync way.

```javascript
export function handleError(error, el, expression = undefined) {
    Object.assign( error, { el, expression } )

    console.warn(`Alpine Expression Error: ${error.message}\n\n${ expression ? 'Expression: \"' + expression + '\"\n\n' : '' }`, el)

    setTimeout( () => { throw error }, 0 )
}
```

I think the correct way to rethrow the error in an async context would be:

```javascript
setTimeout( async () => { throw error }, 0 )
```

But I don't know Alpine code enough to submit a PR and unit test it.

Regards,

Fabiano

**Comment by ekwoka** [comment]
As mentioned by @SimoTod, due to how Alpine handles evaluating the expressions, it wraps the expression to allow it to (hopefully) run it safely without crashing the entirety of Alpine on any one expression being bad, as well as to point you to the expression causing the issue.

I think, in this case, you might also get a different result if you remove the `()` from the `@click` since then Alpine will use the expression to only get a reference to the function and then call it directly (similar to the `onclick` as opposed to running the function inside of the safe catch. I could be wrong though.

I'm not entirely sure there would be any benefit in making any changes to Alpine for this purpose. You should be handling the errors closer to where they actually happen, and not at the window level.

**Reply by personalsoft-fabianobonin** [reply]
There is nothing wrong with how Alpine handle an error and avoid crashes, this is working as expected.

The issue here is that Alpine is introducing a behavior in asyncronous error handling that is distinct than the expected behavior in a javascript environment.

The statement you did about handling errors closer is not a rule and does not solve the problem even.

Let's suppose a have an async method in a frontend application that can throw an exception, but even it i catch it I want to rethrow it to be handled by a global handler to show a notification for the user, for example... There is nothing wrong or unusual with that.

Regards,

Fabiano

**Reply by ekwoka** [reply]
I'm just not sure why, if you're catching it lower, it wouldn't just be turned into a more normal event.

But also, if you remove the `()` it will just do the sync error.

As far as Alpine is concerned, it can't easily differentiate between a sync or async call, since expressions are evaluated as async functions

**Reply by personalsoft-fabianobonin** [reply]
About removing the (), it doesn't work. Same behavior as before. Even if it did work, it would be useless if I have to pass function arguments.

If expressions are always handled asynchronously by Alpine, maybe it must always re-throw it asynchronously, also.

I don't know what you call a "normal event". For me, window events and exception handling are normal events.

Usually I catch the exceptions closer, and re-throw them with a nice, informative and user friendly message, like "Something wrong happened", and let the global handler catch it and show a dialog to the user. This is a pretty normal flow and use case for me. With the advantage that eventually an unhandled event will be handled also.

I think as Alpine is a tool supposed to be used by thousand or even millions of users in different environments, we can't define what is "normal", although I am sure what I am doing is absolutely normal.

I think Alpine must follow javascript specification, and the specification is clear about which event must be send when a error is not handled in an asynchronous context, and it is a "unhandledrejection" event, not an "error" event.

Regards,

Fabiano

**Reply by ekwoka** [reply]
Technically the error is being handled. Whats stranger is just that it's doing both, but I guess that is what the browser is choosing to do when an unhandledrejection error is being thrown outside of a promise. That seems like it's more actually undefined behavior.

But yes, normal event as in that you'd send up your toast event or something. Throwing errors is not meant to be used as an alternative to the event system.

Interestingly, in my local dev sandbox I have for working on Alpine, both of the buttons just do the `error` handler, and not the unhandledrejection at all. quite strange.

Also, the specification is not clear that this should be an unhandled rejection event, because it is, in fact, handled

**Reply by personalsoft-fabianobonin** [reply]
> Technically the error is being handled. Whats stranger is just that it's doing both, but I guess that is what the browser is choosing to do when an unhandledrejection error is being thrown outside of a promise. That seems like it's more actually undefined behavior.

No, it is not. It is being catched by Alpine and re-throwed. This is what is called an unhandled exception.

> But yes, normal event as in that you'd send up your toast event or something. Throwing errors is not meant to be used as an alternative to the event system.

I think you are misunderstanding what an exception is used for, and what is exception handling at all.

> Interestingly, in my local dev sandbox I have for working on Alpine, both of the buttons just do the `error` handler, and not the unhandledrejection at all. quite strange.
>
> Also, the specification is not clear that this should be an unhandled rejection event, because it is, in fact, handled

If you catch an exception and throw another one, the new exception you threw is unhandled. It is what Alpine is doing.

The specification is clear as water, as you can see here:

https://developer.mozilla.org/en-US/docs/Web/API/Window/unhandledrejection_event

**Reply by ekwoka** [reply]
Read that again.

It's unhandled rejection.

The rejection is handled.

It's no longer in a promise.

So its not an unhandled promise rejection.

It's just a normal error.

It's not an unhandled error. It's an unhandled promise rejection. That's what the spec is.

**Reply by personalsoft-fabianobonin** [reply]
Please review the specifications again.

When an error occurs inside a promise, it must be converted into an unhandled promise event. Your attempt to use semantics to justify your perspective is not productive.

The reality is that Alpine's behavior is incorrect, regardless of your argument. If an error occurs within an async method and is not handled, it must be reported as an unhandled promise event. Period.

If Alpine attempts to handle the error and rethrow it as if it were thrown by a synchronous method, it is not following the correct protocol.

If you are only interested in proving your point of view, I kindly ask you to refrain from participating in this discussion. Others may perceive this as a genuine effort to discuss and propose solutions to the problem, which is not the case.

**Reply by SimoTod** [reply]
The thing is that every directive in alpine is wrapped in an asynchronous function even if the user doesn't know. This is to allow users to use await in their directives.
It's a specific design choice to rethrow an error, I don't think it's really trying to match the specs, just to make the dev experience nicer.
I would say that having 2 events is a bug though. The project is open source so feel free to give it a go if you can see a good solution.

**Reply by personalsoft-fabianobonin** [reply]
If Alpine is configured to run everything asynchronously, it may seem like the solution to error handling is as simple as always re-throwing the error asynchronously.

In `error.js`, `handleError` function:

```javascript
setTimeout( async () => { throw error }, 0 )
```

In the proposed test case, both button3 and button4 errors would be handled globally by an "unhandledrejection" event listener, which is the expected behavior.

Can I submit a PR?

Regards,

Fabiano

**Reply by SimoTod** [reply]
I believe the majority of the devs will expect a normal error. The fact that functions are wrapped in an async one is not really documented, it's just an internal design choice. Throwing an unhandledrejection all the time even when the function seems to be synchronous will definitely confuse more developers.
I can't really tell you if the PR would be accepted: I suspect that, without a concrete example where this behavior is a problem, it will be ignored but you can try.

**Reply by personalsoft-fabianobonin** [reply]
Look, I worked around this problem in my code just before I posted this issue here, this is not a problem for me.

I am discussing it here just because I think other developers will face this same problem as this behavior is unexpected and it is not easy to find the cause, so it would be nice to avoid further headaches for our coleagues if possible.

I think there is no problem at all to receive an "error" event instead of an "unhandledrejection" event. The problem is to receive **both**.

I am looking closer at the code, and it seems the async error is leaking somehow when it shouldn't.

First I am looking at `generateEvaluatorFromString` function, and I am trying to figure out in line 103 `if (typeof func === 'function' )`, in what case this would not be a function, as it was generated by `generateFunctionFromString` and it seems the result is always a function?

Also, in line 107 `if (func.finished)`, in what case this function can run synchronously? It seems `generateFunctionFromString` always return an async function.

I am not used with Alpine code and I don't think I have enough skills in JS to handle this problem, but I am trying to understand the code to be able at least to find a way to not generate the "unhandledrejection" event.

Fabiano

**Reply by SimoTod** [reply]
I appreciate that, I'm not trying to talk you down. I was just warning you that the first proposed solution would probably be rejected (no pun intended :) ).
I don't think it can ever run synchronously but I suspect the meaning of that is to cover the case where the promise is already resolved. It's just a guess though.
Re unhandled promises, this one could be a suspect: https://github.com/alpinejs/alpine/blob/6142a83956685b4ddf67d95cf70a46fa3efe5fdf/packages/alpinejs/src/evaluator.js#L136

**Reply by ekwoka** [reply]
> First I am looking at generateEvaluatorFromString function, and I am trying to figure out in line 103 if (typeof func === 'function' ), in what case this would not be a function, as it was generated by generateFunctionFromString and it seems the result is always a function?

`func` can actually be a `Promise<void>` here.

To see this with types you can check https://github.com/ekwoka/Alpine-TS/blob/main/packages/alpinejs/src/evaluator.ts

I'm not exactly sure why that was the way it was chosen...but....it was.

**Reply by personalsoft-fabianobonin** [reply]
It seems the problem is in `evaluator.js`, line 104.

```javascript
let promise = func(func, completeScope).catch((error) => handleError(error, el, expression))
```

The first expression `func(func, completeScope)` is leaking a "unhandledrejection" event before it is caught by `.catch(...)`.

Fabiano
